// react-background-animation/dist/cssEffects/seaWave/seaWave.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _lodash = _interopRequireDefault(require("lodash"));

require("./seaWave.scss");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var SeaWaveEffect =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(SeaWaveEffect, _PureComponent);

  function SeaWaveEffect() {
    _classCallCheck(this, SeaWaveEffect);

    return _possibleConstructorReturn(this, _getPrototypeOf(SeaWaveEffect).apply(this, arguments));
  }

  _createClass(SeaWaveEffect, [{
    key: "render",
    value: function render() {
      var _this$props$className = this.props.className,
          className = _this$props$className === void 0 ? '' : _this$props$className;
      return _react["default"].createElement("div", {
        className: className
      }, _react["default"].createElement("svg", {
        version: "1.1",
        xmlns: "http://www.w3.org/2000/svg",
        xmlnsXlink: "http://www.w3.org/1999/xlink",
        x: "0px",
        y: "0px",
        width: "100vw",
        height: "150px",
        viewBox: "0 0 800 350",
        style: {
          enableBackground: 'new 0 0 800 400',
          left: 0,
          bottom: 0,
          position: 'fixed'
        },
        xmlSpace: "preserve"
      }, _react["default"].createElement("style", {
        type: "text/css"
      }, ".st0{opacity:0.76;}\n          .st1{fill:url(#SVGID_1_);width:100%;}\n          .st2{fill:url(#SVGID_2_);width:100%;}\n          .st3{fill:url(#SVGID_3_);width:100%;}\n          .st4{fill:none;stroke:#FFFFFF;stroke-width:4;stroke-miterlimit:10;width:100%;}"), _react["default"].createElement("g", {
        id: "DarkWaves",
        className: "DarkWaves"
      }, _react["default"].createElement("g", null, _react["default"].createElement("linearGradient", {
        id: "SVGID_2_",
        gradientUnits: "userSpaceOnUse",
        x1: "400.0005",
        y1: "602",
        x2: "400.0005",
        y2: "202.3545"
      }, _react["default"].createElement("stop", {
        offset: "0",
        style: {
          stopColor: '#0000FF'
        }
      }), _react["default"].createElement("stop", {
        offset: "0.1789",
        style: {
          stopColor: '#0020EC'
        }
      }), _react["default"].createElement("stop", {
        offset: "0.4149",
        style: {
          stopColor: '#0043D7'
        }
      }), _react["default"].createElement("stop", {
        offset: "0.6374",
        style: {
          stopColor: '#005CC8'
        }
      }), _react["default"].createElement("stop", {
        offset: "0.8386",
        style: {
          stopColor: '#006CBF'
        }
      }), _react["default"].createElement("stop", {
        offset: "1",
        style: {
          stopColor: '#0071BC'
        }
      })), _react["default"].createElement("path", {
        className: "st2",
        d: "M761.8,202.4c-24.7,0-36.7,7.2-49.5,14.7c-13.2,7.8-26.8,15.9-53.8,15.9c-27,0-40.6-8.1-53.8-15.9\r c-12.8-7.6-24.9-14.7-49.6-14.7c-24.7,0-36.8,7.2-49.6,14.7c-13.2,7.8-26.8,15.9-53.8,15.9c-27,0-40.7-8.1-53.8-15.9\r c-12.8-7.6-24.9-14.7-49.6-14.7c-24.7,0-36.8,7.2-49.6,14.7c-13.2,7.8-26.8,15.9-53.8,15.9c-27,0-40.7-8.1-53.8-15.9\r c-12.8-7.6-24.9-14.7-49.6-14.7c-24.7,0-36.8,7.2-49.6,14.7C78.7,224.9,65,232.9,38,232.9c-27,0-40.4-8.1-53.6-15.9\r c-12.1-7.1-24.3-13.9-45.4-14.6V602h922V232.9c-22.6-0.8-37.2-8.4-49.7-15.8C798.5,209.5,786.5,202.4,761.8,202.4z"
      }))), _react["default"].createElement("g", {
        id: "LightWaves",
        className: "LightWaves"
      }, _react["default"].createElement("g", null, _react["default"].createElement("linearGradient", {
        id: "SVGID_3_",
        gradientUnits: "userSpaceOnUse",
        x1: "400.0005",
        y1: "600",
        x2: "400.0005",
        y2: "214.3545"
      }, _react["default"].createElement("stop", {
        offset: "0",
        style: {
          stopColor: '#0000FF'
        }
      }), _react["default"].createElement("stop", {
        offset: "0.1643",
        style: {
          stopColor: '#0B2CF7'
        }
      }), _react["default"].createElement("stop", {
        offset: "0.3574",
        style: {
          stopColor: '#1559F0'
        }
      }), _react["default"].createElement("stop", {
        offset: "0.5431",
        style: {
          stopColor: '#1E7DEA'
        }
      }), _react["default"].createElement("stop", {
        offset: "0.7168",
        style: {
          stopColor: '#2496E6'
        }
      }), _react["default"].createElement("stop", {
        offset: "0.874",
        style: {
          stopColor: '#28A6E3'
        }
      }), _react["default"].createElement("stop", {
        offset: "1",
        style: {
          stopColor: '#29ABE2'
        }
      })), _react["default"].createElement("path", {
        className: "st3",
        d: "M750.9,229.8c-14.8-7.9-28.7-15.4-57.2-15.4c-28.5,0-42.4,7.5-57.2,15.4c-15.2,8.2-30.9,16.6-62.1,16.6\r s-46.9-8.4-62.1-16.6c-14.8-7.9-28.7-15.4-57.2-15.4c-28.5,0-42.4,7.5-57.2,15.4c-15.2,8.2-30.9,16.6-62.1,16.6\r c-31.2,0-46.9-8.4-62.1-16.6c-14.8-7.9-28.7-15.4-57.2-15.4c-28.5,0-42.4,7.5-57.2,15.4c-15.2,8.2-30.9,16.6-62.1,16.6\r c-31.2,0-46.9-8.4-62.1-16.6c-14.8-7.9-28.9-15.4-57.3-15.4c-16.9,0-28.8,2.6-38.8,6.4V600h922V237c-12,5.3-26,9.4-47.8,9.4\r C782.1,246.4,766.1,237.9,750.9,229.8z"
      }))), _react["default"].createElement("g", {
        id: "WhiteWaves",
        className: "WhiteWaves"
      }, _react["default"].createElement("g", {
        className: "st0"
      }, _react["default"].createElement("path", {
        className: "st4",
        d: "M-71,220c94.2,0,94.2,18,188.4,18c94.2,0,94.2-18,188.4-18c94.2,0,94.2,18,188.4,18c94.2,0,94.2-18,188.4-18\r s94.2,18,188.4,18"
      })))));
    }
  }]);

  return SeaWaveEffect;
}(_react.PureComponent);

var _default = SeaWaveEffect;
exports["default"] = _default;
