// react-background-animation/dist/cssEffects/rain/rain.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _lodash = _interopRequireDefault(require("lodash"));

var _jquery = _interopRequireDefault(require("jquery"));

var _rainStyled = require("./rainStyled");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var RainEffect =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(RainEffect, _PureComponent);

  function RainEffect(props) {
    var _this;

    _classCallCheck(this, RainEffect);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(RainEffect).call(this, props));
    _this.excuteEffect = _this.excuteEffect.bind(_assertThisInitialized(_this));
    _this.makeItRain = _this.makeItRain.bind(_assertThisInitialized(_this));
    return _this;
  }

  _createClass(RainEffect, [{
    key: "makeItRain",
    value: function makeItRain() {
      //clear out everything
      (0, _jquery["default"])('.rain').empty();
      var increment = 0;
      var drops = "";
      var backDrops = "";

      while (increment < 120) {
        //couple random numbers to use for various randomizations
        //random number between 98 and 1
        var randoHundo = Math.floor(Math.random() * (98 - 1 + 1) + 1); //random number between 5 and 2

        var randoFiver = Math.floor(Math.random() * (5 - 2 + 1) + 2); //increment

        increment += randoFiver; //add in a new raindrop with various randomizations to certain CSS properties

        drops += '<div class="drop" style="left: ' + increment + '%; bottom: ' + (randoFiver + randoFiver - 1 + 100) + '%; animation-delay: 0.' + randoHundo + 's; animation-duration: 0.5' + randoHundo + 's;"><div class="stem" style="animation-delay: 0.' + randoHundo + 's; animation-duration: 0.5' + randoHundo + 's;"></div><div class="splat" style="animation-delay: 0.' + randoHundo + 's; animation-duration: 0.5' + randoHundo + 's;"></div></div>';
        backDrops += '<div class="drop" style="right: ' + increment + '%; bottom: ' + (randoFiver + randoFiver - 1 + 100) + '%; animation-delay: 0.' + randoHundo + 's; animation-duration: 0.5' + randoHundo + 's;"><div class="stem" style="animation-delay: 0.' + randoHundo + 's; animation-duration: 0.5' + randoHundo + 's;"></div><div class="splat" style="animation-delay: 0.' + randoHundo + 's; animation-duration: 0.5' + randoHundo + 's;"></div></div>';
      }

      (0, _jquery["default"])('.rain.front-row').append(drops);
      (0, _jquery["default"])('.rain.back-row').append(backDrops);
    }
  }, {
    key: "excuteEffect",
    value: function excuteEffect() {
      this.makeItRain();
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      this.excuteEffect();
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          _this$props$frontRow = _this$props.frontRow,
          frontRow = _this$props$frontRow === void 0 ? true : _this$props$frontRow,
          _this$props$backRow = _this$props.backRow,
          backRow = _this$props$backRow === void 0 ? true : _this$props$backRow,
          _this$props$single = _this$props.single,
          single = _this$props$single === void 0 ? false : _this$props$single,
          className = _this$props.className;
      return _react["default"].createElement("div", {
        className: this.props.className || ''
      }, _react["default"].createElement(_rainStyled.Rain, null, _react["default"].createElement("div", {
        className: "back-row-toggle splat-toggle ".concat(single ? 'single-toggle' : '', " ").concat(className || '')
      }, frontRow && _react["default"].createElement("div", {
        className: "rain front-row"
      }), backRow && _react["default"].createElement("div", {
        className: "rain back-row"
      }))));
    }
  }]);

  return RainEffect;
}(_react.PureComponent);

var _default = RainEffect;
exports["default"] = _default;
